packages/node/src/util/cache.ts

```diff
@@ -0,0 +1,113 @@
+class ListNode<T> {
+  prev: ListNode<T> | null;
+  next: ListNode<T> | null;
+  data: T;
+
+  constructor(data: T) {
+    this.prev = null;
+    this.next = null;
+    this.data = data;
+  }
+}
+
+interface CacheItem<T> {
+  key: string;
+  value: T;
+  createdAt: number;
+}
+
+export class Cache<T> {
+  private readonly capacity: number;
+  private readonly ttlMillis: number;
+  private cache: Map<string, ListNode<CacheItem<T>>>;
+  private head: ListNode<CacheItem<T>> | null;
+  private tail: ListNode<CacheItem<T>> | null;
+
+  constructor(capacity: number, ttlMillis: number) {
+    this.capacity = capacity;
+    this.ttlMillis = ttlMillis;
+    this.cache = new Map();
+    this.head = null;
+    this.tail = null;
+  }
+
+  put(key: string, value: T): void {
+    if (this.cache.has(key)) {
+      this.removeFromList(key);
+    } else if (this.cache.size >= this.capacity) {
+      this.evictLRU();
+    }
+
+    const cacheItem: CacheItem<T> = {
+      key,
+      value,
+      createdAt: Date.now(),
+    };
+
+    const node = new ListNode(cacheItem);
+    this.cache.set(key, node);
+    this.insertToList(node);
+  }
+
+  get(key: string): T | undefined {
+    const node = this.cache.get(key);
+    if (node) {
+      const timeElapsed = Date.now() - node.data.createdAt;
+      if (timeElapsed > this.ttlMillis) {
+        this.remove(key);
+        return undefined;
+      }
+
+      this.removeFromList(key);
+      this.insertToList(node);
+      return node.data.value;
+    }
+    return undefined;
+  }
```

Comment:
In this function we're returing undefined, but in the filter we're checking if the value `== null` which technically works, but I have been burned by null vs undefined before and generally now just prefer using undefined everywhere. 

Microsoft's typescript style guide (microsoft created typescript) says that you should never use `null` explicitly, and always use undefined :)

Reply:
Lets chat about this more broadly, consider this a comment on the entire PR.